src/pykeen/training/callbacks.py

```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
+
+"""Training callbacks.
+
+Training callbacks allow for arbitrary extension of the functionality of the :class:`pykeen.training.TrainingLoop`
+without subclassing it. Each callback instance has a ``loop`` attribute that allows access to the parent training
+loop and all of its attributes, including the model. The interaction points are similar to those of
+`Keras <https://keras.io/guides/writing_your_own_callbacks/#an-overview-of-callback-methods>`_.
+
+Examples
+--------
+It was suggested in `Issue #333 <https://github.com/pykeen/pykeen/issues/333>`_ that it might
+be useful to log all batch losses. This could be accomplished with the following:
+
+.. code-block:: python
+
+    from pykeen.training import TrainingCallback
+
+    class BatchLossReportCallback(TrainingCallback):
+        def on_batch(self, epoch: int, batch_loss: float):
+            print(epoch, batch_loss)
+"""
+
+from typing import Collection, List, Union
+
+from ..trackers import ResultTracker
+
+__all__ = [
+    'TrainingCallbackHint',
+    'TrainingCallback',
+    'TrackerCallback',
+    'MultiTrainingCallback',
+]
+
+
+class TrainingCallback:
+    """An interface for training callbacks."""
+
+    def __init__(self):
+        """Initialize the callback."""
+        self._loop = None
+
+    @property
+    def loop(self):  # noqa:D401
+        """The training loop."""
+        if self._loop is None:
+            raise ValueError('Callback was never initialized')
+        return self._loop
+
+    def register_loop(self, loop) -> None:
+        """Register the training loop."""
+        self._loop = loop
+
+    def on_batch(self, epoch: int, batch_loss: float) -> None:
+        """Call for training batches."""
+
+    def post_batches(self, epoch: int, batch) -> None:
```

Comment:
```suggestion
    def post_batch(self, epoch: int, batch) -> None:
```
for consistency

Reply:
Good suggestion, thanks. Done in 0d6b8005